guppylang/prelude/_internal.py

```diff
@@ -145,16 +122,12 @@ def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
 
         for i in range(len(args)):
             args[i], ty = ExprSynthesizer(self.ctx).synthesize(args[i])
-            if isinstance(ty, OpaqueType) and ty.defn == self.ctx.globals["int"]:
+            if isinstance(ty, NumericType) and ty.kind == NumericType.Kind.Int:
```

Comment:
Shouldn't `nat` be valid too? Maybe bool?

Reply:
arguably anything that implements `__float__`...

Reply:
Yeah, we need a better way to figure out which types are automatically coreced (also applies to function tuples etc.). These `CustomCallChecker`s are quite hacky...

I'm tempted to just leave it as is until we have a proper solution?

Reply:
> arguably anything that implements __float__...

This only works as long as float is at the top of the numeric hierarchy. For example, if we add `complex` in the future, it might also have a `__float__` method but shouldn't be coreced automtically.

Also we'd need coercsion from `nat` to `int` etc...

Reply:
I think the better interim solution is that we coerce `NumericType` instead of just `int`, then come up with something better down the line